python_tsp/heuristics/lin_kernighan.py

```diff
@@ -0,0 +1,263 @@
+            hamiltonian_path_distance
+            + distance_matrix[b][d]
+            - distance_matrix[c][d]
+            + path_cb_distance
+            - path_bc_distance
+        )
+
+        if (
+            new_hamiltonian_path_distance_found + distance_matrix[a][c]
+            < hamiltonian_cycle_distance
+        ):
+            return c, d, new_hamiltonian_path_distance_found
+
+        if (
+            tabu[c][d] != iteration
+            and new_hamiltonian_path_distance_found
+            < hamiltonian_path_distance_found
+        ):
+            hamiltonian_path_distance_found = (
+                new_hamiltonian_path_distance_found
+            )
+            best_c = c
+
+        last_c = c
+        c = d
+
+    return best_c, successors[best_c], hamiltonian_path_distance_found
+
+
+def _print_message(
+    msg: str, verbose: bool, log_file_handler: Optional[TextIO]
+) -> None:
+    if log_file_handler:
+        print(msg, file=log_file_handler)
+
+    if verbose:
+        print(msg)
+
+
+def solve_tsp_lin_kernighan(
+    distance_matrix: np.ndarray,
+    x0: Optional[List[int]] = None,
+    log_file: Optional[str] = None,
+    verbose: bool = False,
+) -> Tuple[List[int], float]:
+    """
+    Solve the Traveling Salesperson Problem using the Lin-Kernighan algorithm.
+
+    Parameters
+    ----------
+    distance_matrix
+        Distance matrix of shape (n x n) with the (i, j) entry indicating the
+        distance from node i to j
+
+    x0
+        Initial permutation. If not provided, it starts with a random path.
+
+    log_file
+        If not `None`, creates a log file with details about the whole
+        execution.
+
+    verbose
+        If true, prints algorithm status every iteration.
+
+    Returns
+    -------
+    Tuple
+        A tuple containing the Hamiltonian cycle and its distance.
+
+    References
+    ----------
+    Éric D. Taillard, "Design of Heuristic Algorithms for Hard Optimization,"
+    Chapter 5, Section 5.3.2.1: Lin-Kernighan Neighborhood, Springer, 2023.
+    """
+    hamiltonian_cycle, hamiltonian_cycle_distance = setup_initial_solution(
+        distance_matrix=distance_matrix, x0=x0
+    )
+    num_vertices = distance_matrix.shape[0]
+    vertices = list(range(num_vertices))
+    iteration = 0
+    improvement = True
+    tabu = np.zeros(shape=(num_vertices, num_vertices), dtype=int)
+
+    log_file_handler = (
+        open(log_file, "w", encoding="utf-8") if log_file else None
+    )
+
+    while improvement:
+        iteration += 1
+        improvement = False
+        successors = _cycle_to_successors(hamiltonian_cycle)
+
+        # Eject edge [a, b] to start the chain and compute the Hamiltonian
+        # path distance obtained by ejecting edge [a, b] from the cycle
+        # as reference.
+        a = int(distance_matrix[vertices, successors].argmax())
+        b = successors[a]
+        hamiltonian_path_distance = (
+            hamiltonian_cycle_distance - distance_matrix[a][b]
```

Comment:
For ND arrays [indexing is better like this](https://numpy.org/doc/stable/user/basics.indexing.html#basic-indexing):
```suggestion
            hamiltonian_cycle_distance - distance_matrix[a, b]
```

When using `distance_matrix[a][b]` you first return a 1-length array in row `a` and then get the element in position `b`, while in `distance_matrix[a, b]` you get the desired element directly.

Reply:
Good catch, can we discuss more on the heuristic LK [PR](https://github.com/fillipe-gsm/python-tsp/pull/39)?

Reply:
@fillipe-gsm I made the suggested indexing improvement and I could see from the tests that the heuristic was even faster! Thanks!